contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -900,8 +895,130 @@ clearly described and the permissions issues should be resolved.
+fooRef:
+    group: sns.services.k8s.aws
+    resource: topics
+    name: foo
+    namespace: foo-namespace
+```
+
+Although not always necessary to help a controller identify a resource type, “group” is included to avoid ambiguity when the resource exists in multiple groups. It also provides clarity to end users and enables copy-pasting of a reference without the referenced type changing due to a different controller handling the reference.
+
+##### Controller behavior
+
+The operator can store a map of (group,resource) to the version of that resource it desires. From there, it can construct the full path to the resource, and retrieve the object.
+
+It is also possible to have the controller choose a version that it finds via the discovery client. However, as schemas can vary across different versions
+of a resource, the controller must also handle these differences.
+
+#### Generic object reference
+
+A generic object reference is used when the desire is to provide a pointer to some object to simplify discovery for the user. For example, this could be used to reference a target object for a `core.v1.Event` that occurred.
+
+With a generic object reference, it is not possible to extract any information about the referenced object aside from what is standard (e.g. ObjectMeta). Since any standard fields exist in any version of a resource, it is possible to not include version in this case:
+
+```yaml
+fooObjectRef:
+    group: operator.openshift.io
+    resource: openshiftapiservers
+    name: cluster
+    # namespace is unset if the resource is cluster-scoped, or lives in the 
+    # same namespace as the referrer.
+```
+
+##### Controller behavior
 
+The operator would be expected to find the resource via the discovery client (as the version is not supplied). As any retrievable field would be common to all objects, any version of the resource should do.
+
+#### Field reference
+
+A field reference is used when the desire is to extract a value from a specific field in a referenced object.
+
+Field references differ from other reference types, as the operator has no knowledge of the object prior to the reference. Since the schema of an object can differ for different versions of a resource, this means that a “version” is required for this type of reference.
+
+```yaml
+fooFieldRef:
+   version: v1 # version of the resource
+   # group is elided in the ConfigMap example, since it has a blank group in the OpenAPI spec.
+   resource: configmaps
+   fieldPath: data.foo
+```
+
+The fieldPath should point to a single value, and use [the recommended field selector notation](#selecting-fields) to denote the field path.
+
+##### Controller behavior
+
+In this scenario, the user will supply all of the required path elements: group, version, resource, name, and possibly namespace.
+As such, the controller can construct the API prefix and query it without the use of the discovery client:
+
+```
+/apis/{group}/{version}/{resource}/
+```
 
```

Comment:
When we talk about controllers taking arbitrary target inputs and doing things to those objects (either writing to them or extracting information from them), it's worth pointing out the security pitfalls to guard against. For example:

* The individual bits of data used to form an API request must be validated to be well-formed (e.g. not be `""` or contain `..` or `/` or escaped variants of those that could expose path traversal vulnerabilities)
* The controller should guard against taking action on the target object (e.g. updating or deleting it) or exposing data fetched from the target object (in events, or log messages, or copying into other API objects) in ways that would let the author of the referencing object escalate permissions or gain visibility to objects they should not have

It's fine if that gets added in a follow-up, this doc has survived this long without it and I'd rather merge bits that have consensus and iterate.

Reply:
Here's a follow-up ticket: https://github.com/kubernetes/community/issues/5797.

I'd like to do that in a follow-up PR as you suggested. Just to make sure that we cover the security component fully and get review on that component independently.